### PR TITLE
Fund Raw Tranaction

### PR DESCRIPTION
**Pull request by dangershony**

### CODE REVIEW

**Comment by bokobza** [comment]
I'll take care of this.

File: Stratis.Bitcoin/Features/Wallet/IWalletManager.cs [hunk 1]
```diff
@@ -226,5 +226,12 @@ public interface IWalletManager : IDisposable
         /// <param name="walletName">The name of the wallet to get.</param>
         /// <returns>A wallet or null if it doesn't exist</returns>
         Wallet GetWalletByName(string walletName);
+
+        /// <summary>
+        /// Get the a change address or create one if all change are used.
```
Comment (bokobza) [inline Stratis.Bitcoin/Features/Wallet/IWalletManager.cs]: "Gets a change address or create one if all change addresses are used."

File: Stratis.Bitcoin/Features/Wallet/IWalletManager.cs [hunk 2]
```diff
@@ -226,5 +226,12 @@ public interface IWalletManager : IDisposable
         /// <param name="walletName">The name of the wallet to get.</param>
         /// <returns>A wallet or null if it doesn't exist</returns>
         Wallet GetWalletByName(string walletName);
+
+        /// <summary>
+        /// Get the a change address or create one if all change are used. 
+        /// </summary>
+        /// <param name="account">The account to create the change on.</param>
```
Comment (bokobza) [inline Stratis.Bitcoin/Features/Wallet/IWalletManager.cs]: "change address"

File: Stratis.Bitcoin/Features/Wallet/WalletTransactionBuilder.cs [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+        /// <param name="context">The context that is used to build a new transaction.</param>
+        /// <returns>The new transaction.</returns>
+        Transaction BuildTransaction(TransactionBuildContext context);
+    }
+
+    /// <summary>
+    /// A builder that uses various parameters to build a Bitcoin transaction.
+    /// </summary>
+    /// <remarks>
+    /// This will uses the FeeEstimator and the TrasnactionBuilder.
```
Comment (bokobza) [inline Stratis.Bitcoin/Features/Wallet/WalletTransactionBuilder.cs]: TrasnactionBuilder -> TransactionBuilder

File: Stratis.Bitcoin/Features/Wallet/WalletTransactionBuilder.cs [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+        private readonly ILogger logger;
+
+        public WalletTransactionBuilder(ILoggerFactory loggerFactory, ConcurrentChain chain, IWalletManager walletManager, IWalletFeePolicy walletFeePolicy, Network network)
+        {
+            this.chain = chain;
+            this.walletManager = walletManager;
+            this.walletFeePolicy = walletFeePolicy;
+            this.network = network;
+            this.coinType = (CoinType)network.Consensus.CoinType;
+            this.logger = loggerFactory.CreateLogger(this.GetType());
```
Comment (bokobza) [inline Stratis.Bitcoin/Features/Wallet/WalletTransactionBuilder.cs]: please use this.logger = loggerFactory.CreateLogger(this.GetType().FullName);

File: Stratis.Bitcoin/Features/Wallet/WalletTransactionBuilder.cs [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+                this.logger.LogError($"Build transaction failed: {string.Join(" - ", errors.Select(s => s.ToString()))}");
+
+                throw new WalletException("Could not build transaction, please make sure you entered the correct data.");
+            }
+
+            return context.Transaction;
+        }
+
+        /// <summary>
+        /// Load all the private keys for each of the <see cref="HdAddress"/> in <see cref="TransactionBuildContext.UnspentOutputs"/>
```
Comment (bokobza) [inline Stratis.Bitcoin/Features/Wallet/WalletTransactionBuilder.cs]: Load's'

File: Stratis.Bitcoin/Features/Wallet/WalletTransactionBuilder.cs [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+        /// </summary>
+        public WalletAccountReference AccountReference { get; set; }
+
+        /// <summary>
+        /// The recipients to send Bitcoin to.
+        /// </summary>
+        public List<Recipient> Recipients { get; set; }
+
+        /// <summary>
+        /// Helper to estimate how much fee to spend on a transaction.
```
Comment (bokobza) [inline Stratis.Bitcoin/Features/Wallet/WalletTransactionBuilder.cs]: is FeeType really a helper?

File: Stratis.Bitcoin/Features/Wallet/WalletTransactionBuilder.cs [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+        }
+
+        /// <summary>
+        /// Add recipients to the <see cref="TransactionBuilder"/>.
+        /// </summary>
+        /// <param name="context">The context associated with the current transaction being built.</param>
+        /// <remarks>
+        /// Add outputs to the <see cref="TransactionBuilder"/> based on the <see cref="Recipient"/> list.
+        /// </remarks>
+        private void AddRecepients(TransactionBuildContext context)
```
Comment (bokobza) [inline Stratis.Bitcoin/Features/Wallet/WalletTransactionBuilder.cs]: should be Rec'i'pients

File: Stratis.Bitcoin/Features/Wallet/WalletManager.cs [hunk 8]
```diff
@@ -168,7 +177,8 @@ public Wallet LoadWallet(string password, string name)
         }
 
         /// <inheritdoc />
-        public Wallet RecoverWallet(string password, string name, string mnemonic, DateTime creationTime, string passphrase = null)
+        public Wallet RecoverWallet(string password, string name, string mnemonic, DateTime creationTime,
```
Comment (dangershony) [inline Stratis.Bitcoin/Features/Wallet/WalletManager.cs]: Ah auto styling yeah I will fix

File: Stratis.Bitcoin/Features/Wallet/WalletManager.cs [hunk 9]
```diff
@@ -1193,13 +1206,45 @@ public TransactionFoundEventArgs(Script script, uint256 transactionHash)
...
+        /// The transaction representing the UTXO.
+        /// </summary>
+        public TransactionData Transaction { get; set; }
+    }
+
+
+    /// <summary>
+    /// Represent a high level account container that hold all its <see cref="UnspentOutputReference"/>.
+    /// </summary>
+    public class UnspentAccountReference
```
Comment (dangershony) [inline Stratis.Bitcoin/Features/Wallet/WalletManager.cs]: It does only contain 1 account, all the child Accounts in UnspentOutputReference are just the same account to provide an easy mapping between Transaction => Address => Account

File: Stratis.Bitcoin/Features/Wallet/WalletManager.cs [hunk 10]
```diff
@@ -1264,19 +1263,19 @@ public OutPoint ToOutPoint()
 
 
     /// <summary>
-    /// Represent a high level account container that hold all its <see cref="UnspentOutputReference"/>.
+    /// Represent a high level account container that hold's all its <see cref="UnspentOutputReference"/>.
```
Comment (bokobza) [inline Stratis.Bitcoin/Features/Wallet/WalletManager.cs]: should be represent's' and hold's'

File: Stratis.Bitcoin/Features/Wallet/WalletTransactionHandler.cs [hunk 11]
```diff
@@ -145,7 +145,7 @@ public void FundTransaction(TransactionBuildContext context, Transaction transac
         }
 
         /// <summary>
-        /// Load all the private keys for each of the <see cref="HdAddress"/> in <see cref="TransactionBuildContext.UnspentOutputs"/>
+        /// Load's all the private keys for each of the <see cref="HdAddress"/> in <see cref="TransactionBuildContext.UnspentOutputs"/>
```
Comment (bokobza) [inline Stratis.Bitcoin/Features/Wallet/WalletTransactionHandler.cs]: Loads doesn't need the apostrophe.

**Pull request merged**